### ISSUE TITLE
Gasses should be stored at the TOP of tanks rather than the bottom

### ISSUE THREAD

**Issue by AlexIIL** [AlexIIL/Buildcraft_1.8_Issues#122]

**Comment by LuigiHutch** [comment]
Just curious, which mods add gases and why would they travel in fluid pipes?

**Comment by AlexIIL** [comment]
no idea

**Comment by AlexIIL** [comment]
you never saw this

**Comment by AlexIIL** [comment]
this is not a hint

**Comment by AlexIIL** [comment]
This has been marked as "invalid" and "not buildcraft" becasue it is definitaly nothing todo with ~~neptune~~BuIldCraft. nothing at all. move along now. nothing to see here.

**Comment by LuigiHutch** [comment]
Heh, ok then :)

**Comment by dj3520** [comment]
> Just curious, which mods add gases and why would they travel in fluid pipes?

Galacticraft does but I don't think they're focused on 1.8 at the moment.

**Comment by AlexIIL** [comment]
Yes! That was totally the reason for this issue! Nothing bigger going on here, no sir!

**Comment by LuigiHutch** [comment]
You're such a tease :3

**Comment by dj3520** [comment]
How? He said nothing else is going on. One should assume that's true. :smile:

**Comment by AlexIIL** [comment]
I'm almost surprised none of you have tried cloning the main repo to see if refinery stuff is there... oh wait its not. nothing wrong there.

**Comment by meelock** [comment]
You just inspired me to make a mod all about gases and extra refinery uses... yey

**Comment by MrGraphic** [comment]
Hehe, nice one @AlexIIL . It sounds like this fancy "refining" system finally comes to live :smiley:

**Comment by MoreThanHidden** [comment]
Back to the main topic, gasses should basically expand to fill the tank not necessarily top or bottom?
![gas](https://cloud.githubusercontent.com/assets/1351964/13105609/ff826972-d59e-11e5-8d83-b703ebe9c74d.gif)

**Comment by dj3520** [comment]
Depends, how hard would it be to change the translucency of a
"fluid-but-really-gas?" If it isn't terribly hard/glitchy then I'd like it.

**Comment by AlexIIL** [comment]
Not difficult at all, just a bit of multiplication. :)

**Comment by AlexIIL** [comment]
Turns out implementing a new refinery system isn't taking anywhere near as long as I though it would. Hopefully I will have something to playtest soon :)

**Comment by dj3520** [comment]
I'm interested in any refinery changes. Nothing wrong with the current system imo but interested in what could be done to it.

**Comment by LuigiHutch** [comment]
My personal thoughts on the current refinery system is that at the moment all you have to do is plug in and power. No thought goes into it and thus makes it not very interesting :/

Hopefully, the new system makes it more interesting :)

**Comment by AlexIIL** [comment]
Its definitely different from before. And its more interesting than just a giant multi-block :)

**Comment by LunNova** [comment]
> I'm almost surprised none of you have tried cloning the main repo to see if refinery stuff is there... oh wait its not. nothing wrong there.

So that's why all my jenkins builds have been failing with missing files for days... d'oh :(

Might be best to put this stuff on a branch like `feature/absolutely-nothing-to-do-with-gases-here` so `master` isn't in a broken state?

**Comment by AlexIIL** [comment]
Yeah, braking the main branch on purpose is a bit of a silly idea. Fortunately I'm nearly done though so it won't be like this for too much longer. (I just need temp. models)

**Comment by dj3520** [comment]
Very interested in making this somewhat more complicated. Not sure how far things are going to go, but there's some info I dug up on [what's known as the Flare Stack](https://www.quora.com/In-refineries-why-do-they-have-a-constantly-burning-flame-at-the-top-of-some-tower-What-purpose-does-it-serve) (the visible flame found at refineries)

From a picture in the comments it looks as though there are several different things you can get from just oil. Maybe diesel engines? Maybe production of kerosine? (Galacticraft has a "body heat" mechanic for traveling to mars. Could use kerosine heaters.)

**Comment by dj3520** [comment]
Perhaps I should mention my grandfather is a retired from the refinery. I think he was a bit higher in command than a supervisor but what exactly his title was I don't think I've ever heard.

**Comment by kotoroshinoto** [comment]
I know this is a mod for a game, but just to put scientific 2 cents in, the behavior of a gas in this sort of situation depends on the properties of that gas AND what conditions exist in the transport pipe.

Even if BC itself does not plan to add gases, it should recognize that mods using its API may choose to do so and come to a sane decision and recommendation to give modders who would use its API as a basis for the use and utility of their own mod's materials. (steam is a fairly common one If I recall correctly)

Technically gas does exhibit many of the properties a typical fluid has, and often is modeled using similar systems. (air turbulence is not unlike fluid turbulence) it obviously  has traits that differ from a canonical fluid that alter its behavior.

If the gas is simply being allowed to flow where it will due to a low feed rate, high pipe volume, low pressure etc then the properties of the gas are fairly important. Some gases that behave more like an ideal gas will behave as described in morethanhidden's image, however if there is already gas present in the pipe (like normal atmosphere), some gases do not mix readily due to having significant difference in density or relatively low volatility. (and will thus either pool at the ceiling or ground level). 

In the event that transfer is considered to be significantly pressurized, then the fill-pipe model is the one to go with.

Short answer is that IF you were going to plan to allow BC fluids to be designated as gases and reverse the animation top/bottom-wise and or even provide for built-in reversed gravity flow direction for the fluid outside of a pipe through a buildcraft base class or API element, then the coder would need the ability to tell buildcraft both that the fluid in question is a gas, which pipe behavior it ought to display, and which flow direction it would take.

sensible list of pipe behavior: pool at top, pool at bottom, spread-to-fill (presumably would darken or lighten  texture colors based on amount present)

I don't have any special insight into what you've already coded, i'm just writing this off the cuff based on what I recall from undergraduate physics, chemistry, and physical chemistry.

**Comment by dj3520** [comment]
@kotoroshinoto The unfortunate part sometimes about realism is to the average player it's a pain to deal with. The idea of having a fluid pipe not allowing gasses across due to air inside the pipe would irritate most players, probably including me. It'd probably be best to assume that all pipes don't contain air and can only contain one gas at a time, which is how fluids currently work. With air in the pipes a fluid couldn't take all of the space inside, yet they do. Though this could be that the pipes are waterproof but not airtight. Either way, unless a pressure system is intentionally added, I feel like pipes shouldn't have any positive or negative pressure. Gas would flow as liquid already does, except possibly upside down or of varying translucency.

Personally I think the different translucencies would be interesting, though I'm concerned how I'd see what gas in a tank when there is a small amount in a large tank (nearly transparent.)

Also there are gases heavier than air that would make more sense to keep near the bottom of the tank. Perhaps this is a more complicated idea, but maybe gasses could have a -1 to 1 floating point variable on how "light" they are. -1 renders from the top first, then diffuses more in the "empty" portion of the tank when closer to 0, creating a gradient. 0 renders as more gas = less transparent. 1 is like a fluid, and again something like 0.5 would have the darkest portion at the bottom and slowly taper off as the tank goes up. So gasses lighter than air would have a negative bias, gasses near air's density would be near 0, and gases heavier would be on the plus side.

This might have neat implications not only with scientific gases but possibly even magic ones. Since I don't mod for minecraft I don't know how hard this would be, so sorry if I'm leaning on the impossible.

**Comment by AlexIIL** [comment]
Forge (not BC) implements Fluid API's, which means that every mod that adds fluids and wants BC to use different values of flow rate can by registering fluids with different values. Forge has a lot of these definitions built in, and as such I can't add any more or change the definitions of them. Fluids in this way expose the following properties:

_(Copied from `Fluid.java`)_

``` java
    /** Density of the fluid - completely arbitrary; negative density indicates that the fluid is
     * lighter than air.
     * Default value is approximately the real-life density of water in kg/m^3. */
    protected int density = 1000;

    /** Temperature of the fluid - completely arbitrary; higher temperature indicates that the fluid is
     * hotter than air.
     * Default value is approximately the real-life room temperature of water in degrees Kelvin. */
    protected int temperature = 300;

    /** Viscosity ("thickness") of the fluid - completely arbitrary; negative values are not
     * permissible.
     * Default value is approximately the real-life density of water in m/s^2 (x10^-3).
     * Higher viscosity means that a fluid flows more slowly, like molasses.
     * Lower viscosity means that a fluid flows more quickly, like helium. */
    protected int viscosity = 1000;

    /** This indicates if the fluid is gaseous.
     * Useful for rendering the fluid in containers and the world.
     * Generally this is associated with negative density fluids. */
    protected boolean isGaseous;
```

Because of how icky translucency is I don't want to implement spread-to-fill, and its possible for mods to have fluids that are very close in colour so lightening the texture would just lead to confusion between different fluids.

> Also there are gases heavier than air that would make more sense to keep near the bottom of the tank. Perhaps this is a more complicated idea, but maybe gasses could have a -1 to 1 floating point variable on how "light" they are. -1 renders from the top first, then diffuses more in the "empty" portion of the tank when closer to 0, creating a gradient. 0 renders as more gas = less transparent. 1 is like a fluid, and again something like 0.5 would have the darkest portion at the bottom and slowly taper off as the tank goes up. So gasses lighter than air would have a negative bias, gasses near air's density would be near 0, and gases heavier would be on the plus side.
> 
> This might have neat implications not only with scientific gases but possibly even magic ones. Since I don't mod for minecraft I don't know how hard this would be, so sorry if I'm leaning on the impossible.

@dj3520 Its definitely possible, however I think that would be very confusing and a big deviation from how fluids display.

**Comment by kotoroshinoto** [comment]
negative density could be used as a flag for fill-top behavior then. If it is stored in a tank of some sort, it should probably fill from the top down, and flow along the top-side of pipes, if that isn't too much of a bear to impliment. (I mean this for lighter than air fluids not necessarily someone's attempt to depict a gas.)
